### ISSUE TITLE
🐛 Bug Report: Backend failing to start in k8s pod when securityContext with readOnlyRootFilesystem: true

### ISSUE THREAD

**Issue by suganselvaraj** [backstage/backstage#26753]
### 📜 Description

We have Styra policies enabled as part of the K8S cluster. One of the policy denies starting any pod container if the pod manifest states that "readOnlyRootFilesystem: false"   When the policy is enforced in the k8s cluster where the backstage pod is running the deployment starts failing in argo . Upon looking at the POD status in the cluster, we found out that the POD was on a crashback loop

### 👍 Expected behavior

Backstage K8S Pod should start without any readOnlyRootFileSystem mount  issue.

### 👎 Actual Behavior with Screenshots

Based on this  [documentation](https://backstage.io/docs/reference/plugin-app-backend.routeroptions.disableconfiginjection/)  from backstage, we tried providing the disableConfigInjection: true    in the app-config.yaml file with an assumption that it will disable the config injection during  runtime but that didn’t work out

`Backend failed to start up ForwardedError: Plugin 'app' startup failed; caused by Error: EROFS: read-only file system, open '/app/packages/app/dist/static/module-backstage.a5db0f1c.js'
    at /app/node_modules/@backstage/backend-app-api/dist/index.cjs.js:1680:19
    at async /app/node_modules/@backstage/backend-app-api/dist/index.cjs.js:1679:11
    at async Promise.all (index 1)
    at async #doStart (/app/node_modules/@backstage/backend-app-api/dist/index.cjs.js:1633:5)
    at async BackendInitializer.start (/app/node_modules/@backstage/backend-app-api/dist/index.cjs.js:1562:5)
    at async BackstageBackend.start (/app/node_modules/@backstage/backend-app-api/dist/index.cjs.js:1762:5) {
  cause: [Error: EROFS: read-only file system, open '/app/packages/app/dist/static/module-backstage.a5db0f1c.js'] {
    errno: -30,
    code: 'EROFS',
    syscall: 'open',
    path: '/app/packages/app/dist/static/module-backstage.a5db0f1c.js'
  }
}
`

### 👟 Reproduction steps

We tried below approach to fix the error ;

Volume Mount:
We were able to fix the error with yarn writing cache to /tmp/.yarn-cache by mounting a volume mount to the /tmp folder .
As per the fix  we also tried to follow the same approach by  trying a volume mount to /app path to a temp location where the backstage is trying to write the file . 
This approach didn’t work out as expected and became tricky as mounting /app created other side effects where the mount volume override the existing file system with the OS along with build artifacts  in /app path in the filesystem.  

Upon creating the volume the POD could not even start with error as below.
                        
                           Could not find entry.sh file.

Based on the above error we removed the volume mount.

Backstage Configuration
Based on this  [documentation](https://backstage.io/docs/reference/plugin-app-backend.routeroptions.disableconfiginjection/)  from backstage, we tried providing the disableConfigInjection: true    in the app-config.yaml file with an assumption that it will disable the config injection during  runtime but that didn’t work out and we were receiving the same error as provided in the stack trace above.

Mount /app as a writable emptyDir in docker image
We tried to copy the contents of the backstage packages bundle which is currently read only /app in the file system to a temporary path during build before copying it to a writable /app during runtime without success.

Other users have reported this error without a full resolution: 
https://github.com/backstage/backstage/issues/25573
https://github.com/backstage/backstage/issues/4135 

The preliminary approach suggested during our initial call with Spotify is to also separate the front-end and backend while mounting the configuration from a writable mount: https://github.com/backstage/backstage/tree/master/contrib/docker/frontend-with-nginx

### 📃 Provide the context for the Bug.

Start the backstage pod with the readOnlyrootFileSystem set to true. But getting below error.

`Backend failed to start up ForwardedError: Plugin 'app' startup failed; caused by Error: EROFS: read-only file system, open '/app/packages/app/dist/static/module-backstage.a5db0f1c.js'
    at /app/node_modules/@backstage/backend-app-api/dist/index.cjs.js:1680:19
    at async /app/node_modules/@backstage/backend-app-api/dist/index.cjs.js:1679:11
    at async Promise.all (index 1)
    at async #doStart (/app/node_modules/@backstage/backend-app-api/dist/index.cjs.js:1633:5)
    at async BackendInitializer.start (/app/node_modules/@backstage/backend-app-api/dist/index.cjs.js:1562:5)
    at async BackstageBackend.start (/app/node_modules/@backstage/backend-app-api/dist/index.cjs.js:1762:5) {
  cause: [Error: EROFS: read-only file system, open '/app/packages/app/dist/static/module-backstage.a5db0f1c.js'] {
    errno: -30,
    code: 'EROFS',
    syscall: 'open',
    path: '/app/packages/app/dist/static/module-backstage.a5db0f1c.js'
  }
}
`

### 🖥️ Your Environment

OS:   Darwin 23.6.0 - darwin/x64
node: v20.8.0
yarn: 1.22.19
cli:  0.26.11 (installed)
backstage:  1.28.2

Dependencies:
  @backstage/app-defaults                                          1.5.8, 1.5.9
  @backstage/backend-app-api                                       0.7.9, 0.8.0, 0.9.3
  @backstage/backend-common                                        0.22.0, 0.23.3, 0.24.1
  @backstage/backend-defaults                                      0.2.18, 0.3.2, 0.4.1, 0.4.4
  @backstage/backend-dev-utils                                     0.1.4, 0.1.5
  @backstage/backend-dynamic-feature-service                       0.2.14
  @backstage/backend-openapi-utils                                 0.1.14
  @backstage/backend-plugin-api                                    0.6.21, 0.6.20, 0.7.0, 0.8.1
  @backstage/backend-tasks                                         0.5.25, 0.5.27
  @backstage/backend-test-utils                                    0.4.2, 0.4.4, 0.5.1
  @backstage/catalog-client                                        1.6.5, 1.6.6
  @backstage/catalog-model                                         1.5.0, 1.6.0
  @backstage/cli-common                                            0.1.14
  @backstage/cli-node                                              0.2.7
  @backstage/cli                                                   0.26.11
  @backstage/config-loader                                         1.8.1, 1.9.0
  @backstage/config                                                1.2.0
  @backstage/core-app-api                                          1.14.0, 1.14.1
  @backstage/core-compat-api                                       0.2.6, 0.2.7
  @backstage/core-components                                       0.13.10, 0.14.9
  @backstage/core-plugin-api                                       1.9.3
  @backstage/dev-utils                                             1.0.35, 1.0.36
  @backstage/e2e-test-utils                                        0.1.1
  @backstage/errors                                                1.2.4
  @backstage/eslint-plugin                                         0.1.8
  @backstage/frontend-plugin-api                                   0.6.7
  @backstage/integration-aws-node                                  0.1.12
  @backstage/integration-react                                     1.1.29
  @backstage/integration                                           1.13.0, 1.14.0
  @backstage/plugin-app-backend                                    0.3.69
  @backstage/plugin-app-node                                       0.1.21
  @backstage/plugin-auth-backend-module-atlassian-provider         0.2.3
  @backstage/plugin-auth-backend-module-aws-alb-provider           0.1.14
  @backstage/plugin-auth-backend-module-azure-easyauth-provider    0.1.5
  @backstage/plugin-auth-backend-module-bitbucket-provider         0.1.5
  @backstage/plugin-auth-backend-module-cloudflare-access-provider 0.1.5
  @backstage/plugin-auth-backend-module-gcp-iap-provider           0.2.17
  @backstage/plugin-auth-backend-module-github-provider            0.1.19
  @backstage/plugin-auth-backend-module-gitlab-provider            0.1.19
  @backstage/plugin-auth-backend-module-google-provider            0.1.17, 0.1.19
  @backstage/plugin-auth-backend-module-guest-provider             0.1.7
  @backstage/plugin-auth-backend-module-microsoft-provider         0.1.17
  @backstage/plugin-auth-backend-module-oauth2-provider            0.2.3
  @backstage/plugin-auth-backend-module-oauth2-proxy-provider      0.1.15
  @backstage/plugin-auth-backend-module-oidc-provider              0.2.3
  @backstage/plugin-auth-backend-module-okta-provider              0.0.13, 0.0.15
  @backstage/plugin-auth-backend-module-onelogin-provider          0.1.3
  @backstage/plugin-auth-backend                                   0.22.9, 0.22.7
  @backstage/plugin-auth-node                                      0.4.17, 0.4.15, 0.5.1
  @backstage/plugin-auth-react                                     0.1.3
  @backstage/plugin-catalog-backend-module-aws                     0.3.16
  @backstage/plugin-catalog-backend-module-github                  0.6.3
  @backstage/plugin-catalog-backend-module-scaffolder-entity-model 0.1.18
  @backstage/plugin-catalog-backend                                1.23.2
  @backstage/plugin-catalog-common                                 1.0.25
  @backstage/plugin-catalog-graph                                  0.4.6
  @backstage/plugin-catalog-import                                 0.12.0
  @backstage/plugin-catalog-node                                   1.12.4
  @backstage/plugin-catalog-react                                  1.12.2
  @backstage/plugin-catalog                                        1.21.0
  @backstage/plugin-events-backend                                 0.3.8
  @backstage/plugin-events-node                                    0.3.10, 0.3.7, 0.3.8
  @backstage/plugin-home-react                                     0.1.14
  @backstage/plugin-home                                           0.7.5
  @backstage/plugin-kubernetes-backend                             0.18.1
  @backstage/plugin-kubernetes-common                              0.8.0
  @backstage/plugin-kubernetes-node                                0.1.14
  @backstage/plugin-kubernetes-react                               0.4.0
  @backstage/plugin-kubernetes                                     0.11.11
  @backstage/plugin-org                                            0.6.26
  @backstage/plugin-permission-backend                             0.5.46
  @backstage/plugin-permission-common                              0.7.14, 0.8.0, 0.8.1
  @backstage/plugin-permission-node                                0.7.32, 0.8.0, 0.8.2
  @backstage/plugin-permission-react                               0.4.24
  @backstage/plugin-proxy-backend                                  0.5.1
  @backstage/plugin-scaffolder-backend-module-azure                0.1.12
  @backstage/plugin-scaffolder-backend-module-bitbucket-cloud      0.1.10
  @backstage/plugin-scaffolder-backend-module-bitbucket-server     0.1.10
  @backstage/plugin-scaffolder-backend-module-bitbucket            0.2.10
  @backstage/plugin-scaffolder-backend-module-cookiecutter         0.2.46
  @backstage/plugin-scaffolder-backend-module-gerrit               0.1.12
  @backstage/plugin-scaffolder-backend-module-gitea                0.1.10
  @backstage/plugin-scaffolder-backend-module-github               0.3.1
  @backstage/plugin-scaffolder-backend-module-gitlab               0.4.2
  @backstage/plugin-scaffolder-backend                             1.22.10
  @backstage/plugin-scaffolder-common                              1.5.3, 1.5.4, 1.5.5
  @backstage/plugin-scaffolder-node-test-utils                     0.1.11, 0.1.7
  @backstage/plugin-scaffolder-node                                0.4.10, 0.4.6, 0.4.7, 0.4.8
  @backstage/plugin-scaffolder-react                               1.10.0, 1.9.0
  @backstage/plugin-scaffolder                                     1.21.0, 1.23.0
  @backstage/plugin-search-backend-module-catalog                  0.1.28
  @backstage/plugin-search-backend-module-elasticsearch            1.5.3
  @backstage/plugin-search-backend-module-pg                       0.5.30
  @backstage/plugin-search-backend-module-techdocs                 0.1.25
  @backstage/plugin-search-backend-node                            1.2.27
  @backstage/plugin-search-backend                                 1.5.12
  @backstage/plugin-search-common                                  1.2.13
  @backstage/plugin-search-react                                   1.7.12
  @backstage/plugin-search                                         1.4.12
  @backstage/plugin-signals-react                                  0.0.4
  @backstage/plugin-techdocs-module-addons-contrib                 1.1.11
  @backstage/plugin-techdocs-node                                  1.12.6
  @backstage/plugin-techdocs-react                                 1.2.5
  @backstage/plugin-techdocs                                       1.10.6
  @backstage/plugin-user-settings-common                           0.0.1
  @backstage/plugin-user-settings                                  0.8.7
  @backstage/release-manifests                                     0.0.11
  @backstage/test-utils                                            1.5.8, 1.5.9
  @backstage/theme                                                 0.4.4, 0.5.6
  @backstage/types                                                 1.1.1
  @backstage/version-bridge                                        1.0.8

### 👀 Have you spent some time to check if this bug has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

### Are you willing to submit PR?

None

**Comment by GeetaAhuja** [comment]
@awanlin FYI

**Comment by awanlin** [comment]
I've been able to reproduce this doing the following:

1. Run `npx @backstage/create-app@latest`
2. Then run `yarn install --immutable`, `yarn tsc`, and `yarn build:backend --config ../../app-config.yaml --config ../../app-config.production.yaml`
3. Now run `docker image build . -f packages/backend/Dockerfile --tag backstage` to build the image
4. Finally run `docker run -it --read-only -p 7007:7007 backstage`

This will fail with the error above. To resolve it I added:

```diff
  app:
    title: Scaffolded Backstage App
    baseUrl: http://localhost:3000
+   disableConfigInjection: true
```

Then went though the steps above skipping step one. This got me a working Backstage instance with a read only file system.

The impact of using the `disableConfigInjection` flag means that all your config needs to be provided at build time as per this comment: https://github.com/backstage/backstage/blob/6f8d71b74ca53b6a62d0ff6325f7849d3b48a822/plugins/app-backend/config.d.ts#L29-L38

**Comment by Rugvip** [comment]
I've realized the change from #26389 was incorrectly marked as a `patch` change rather than  a `minor` bump. At this point we'd need to rever and re-release as a minor bump to fix that, and I'm a bit worried that's gonna make things even messier.  Will still consider that path forward though depending on what we hear from the community.

Regarding the issue, this is currently working as expected in that if you want to run the app backend with a read-only file system you'd need to disable the injection. I'd say we have a pretty good path forward now though, where it's now fairly simple to enable templating and injection of config by serving a virtual `index.html`. I might work on this in a little while, but in the meantime if anyone wants to pick this up I'd be happy to provide a bit of guidance. I think it would be a much better approach than rewriting the files on disk.

**Comment by dlorych** [comment]
Hi @Rugvip, could you elaborate more on path forward ("virtual `index.html`")?

Is it going to be provided by app-backend plugin? or is it supposed to be delivered by nginx (in a separate frontend setup)?

**Comment by Rugvip** [comment]
@dlorych it would be the app plugin backend implementing it. Essentially it would keep and serve the rendered version of `index.html` from memory. The only complicated part of that is that there's quite a lot of logic baked into this line: https://github.com/backstage/backstage/blob/d336f929d8b287227060d62f848803ab87306229/plugins/app-backend/src/service/router.ts#L320

Essentially we just need a version of `.sendFile` that can provide the data from memory.

**Comment by alex13rivera** [comment]
@awanlin disbaling the injection of variable does not work for me. I have environment/region specific configuration that need to be injected. I cannot build that at build time. What is the option to keep injecting and yet run with rootFilesystem to true?

**Comment by freben** [comment]
https://github.com/backstage/backstage/pull/27689
https://github.com/backstage/backstage/pull/27695

These PRs are meant to close this issue. Once the release is out today, feel free to try it and close if it helped.

**Comment by mathieu-benoit** [comment]
@Rugvip, JFYI and based on the discussions we had at the Backstage Kiosk at KubeCon, I now confirmed that with Backstage 1.33.0 (and associated deps `backstage/plugin-app-backend": 0.4.0`, etc.) it's now working successfully with `readOnlyRootFilesystem: true`. Thanks so much!

**Comment by awanlin** [comment]
Closing as resolved in the `1.33.0` release of Backstage 🚀

**Comment by Rugvip** [comment]
@mathieu-benoit awesome, thanks! 👍
